streamrouter/consul.py

```diff
@@ -1,485 +1,552 @@
+        return self.get_services(
+            lib.srl__consul__get_all_arrow_asns_services,
+            lib.srl__arrow_asns_services__next,
+            lib.srl__arrow_asns_services__free,
+            ArrowAsnsService)
 
     def add_update_callback(self, cb):
-        self.__update_callbacks.add(cb)
+        """
+        Add a given update callback. The callback will be invoked whenever
+        the consul service gets updated.
+        """
+        def callback(err):
+            cb()
+
+        ncb = lib.UPDATE_CALLBACK(callback)
+
+        assert self.raw_ptr is not None
 
-    def remove_update_callback(self, cb):
-        self.__update_callbacks.remove(cb)
+        token = lib.srl__consul__add_update_callback(self.raw_ptr, ncb)
+
+        self.update_callbacks[token] = ncb
+
+        return token
+
+    def remove_update_callback(self, token):
+        """
+        Remove an update callback corresponding to a given token.
+        """
+        assert self.raw_ptr is not None
+
+        lib.srl__consul__remove_update_callback(self.raw_ptr, token)
+
+        try:
+            del self.update_callbacks[token]
+        except KeyError:
+            pass
 
     def add_routing_changed_callback(self, cb):
-        self.__routing_changed_callbacks.add(cb)
+        """
+        Add a given routing-changed callback. The callback will be invoked
+        whenever the routing information changes.
+        """
+        ncb = lib.CHANGE_CALLBACK(cb)
 
-    def remove_routing_changed_callback(self, cb):
-        self.__routing_changed_callbacks.remove(cb)
+        assert self.raw_ptr is not None
 
-    def is_healthy(self):
-        return self.__last_error is None
+        token = lib.srl__consul__add_change_callback(self.raw_ptr, ncb)
 
-    def exception(self):
-        return self.__last_error
+        self.change_callbacks[token] = ncb
 
-    def close(self):
-        if self.__closed:
-            return
+        return token
 
-        self.__closed = True
+    def remove_routing_changed_callback(self, token):
+        """
+        Remove a routing-changed callback corresponding to a given token.
+        """
+        assert self.raw_ptr is not None
 
-        self.__sync_task.cancel()
+        lib.srl__consul__remove_change_callback(self.raw_ptr, token)
 
         try:
-            self.__event_loop.run_until_complete(
-                self.__sync_task)
-        except asyncio.CancelledError:
+            del self.change_callbacks[token]
+        except KeyError:
             pass
+
+    def is_healthy(self):
+        """
+        Check if the consul service is healthy.
+        """
+        assert self.raw_ptr is not None
+
+        return lib.srl__consul__get_status(self.raw_ptr) == lib.STATUS_OK
+
+    def exception(self):
+        """
+        Get the last synchronization exception.
+        """
+        assert self.raw_ptr is not None
+
+        err = get_string(lib.srl__consul__get_error, self.raw_ptr)
+        if err is None:
```

Comment:
You could just return `err and SynchronizationError(err)`, but yes this is more readable when you are not sure how `and` works in python. But I like this way - I used it usually with `or`, but why not with `and`.

Reply:
I know how `and` works in Python :) . As you mentioned, I prefer readability here. The expression with `or` is quite commonly used, however, I'm not sure if that's the case with `and` expressions.